man/man3/memory.3
.\" @(#)memory.3 1.1 92/07/30 SMI; from S5
.TH MEMORY 3 "6 October 1987"
.SH NAME
memory, memccpy, memchr, memcmp, memcpy, memset \- memory operations
.SH SYNOPSIS
.LP
.nf
.ft B
.B #include <memory.h>
.ft
.fi
.LP
.nf
.ft B
char *memccpy(s1, s2, c, n)
char *s1, *s2;
int c, n;
.ft
.fi
.LP
.nf
.ft B
char *memchr(s, c, n)
char *s;
int c, n;
.ft
.fi
.LP
.nf
.ft B
int memcmp(s1, s2, n)
char *s1, *s2;
int n;
.ft
.fi
.LP
.nf
.ft B
char *memcpy(s1, s2, n)
char *s1, *s2;
int n;
.ft
.fi
.LP
.nf
.ft B
char *memset(s, c, n)
char *s;
int c, n;
.ft
.fi
.SH DESCRIPTION
.IX "memccpy function" "" "\fLmemccpy()\fR \(em copy memory character strings"
.IX "copy" "memory character strings \(em \fLmemccpy()\fR"
.IX "memchr()" "" "\fLmemchr()\fR \(em index memory characters"
.IX "index memory characters memchr()" "" "index memory characters \(em \fLmemchr()\fR"
.IX "memcmp()" "" "\fLmemcmp()\fR compare memory characters"
.IX "compare" "memory characters \(em \fLmemcmp()\fR"
.IX "memcpy()" "" "\fLmemcpy()\fR copy memory character fields"
.IX "copy" "memory character fields \(em \fLmemcpy()\fR"
.IX "memset()" "" "\fLmemset()\fR assign to memory characters"
.IX "assign to memory characters memset" "" "assign to memory characters \(em \fLmemset()\fR"
.IX "memory operations"
.LP
These functions operate as efficiently as
possible on memory areas (arrays of characters
bounded by a count, not terminated by a
null
character).  They do not check for the overflow
of any receiving memory area.
.LP
.B memccpy(\|)
copies characters from memory area
.I s2
into
.IR s1 ,
stopping after the first occurrence of character
.I c
has been copied, or after
.I n
characters have been copied, whichever comes first.
It returns a pointer to the character after
the copy of
.I c
in
.IR s1 ,
or a
.SM NULL
pointer if
.I c
was not found in the first
.I n
characters of
.IR s2 .
.LP
.B memchr(\|)
returns a pointer to the first
occurrence of character
.I c
in the first
.I n
characters of memory area
.IR s ,
or a
.SM NULL
pointer if
.I c
does not occur.
.LP
.B memcmp(\|)
compares its arguments, looking at the first
.I n
characters only, and returns an integer
less than, equal to, or greater than 0,
according as
.I s1
is lexicographically less than, equal to, or
greater than
.IR s2 .
.LP
.B memcpy(\|)
copies
.I n
characters from memory area
.I s2
to
.IR s1 .
It returns
.IR s1 .
.LP
.B memset(\|)
sets the first
.I n
characters in memory area
.I s
to the value of character
.IR c .
It returns
.IR s .
.SH NOTES
For user convenience, all these functions are declared in the
.B <memory.h>
header file.
.SH BUGS
.B memcmp(\|)
uses native character comparison, which
is signed on some machines
and unsigned on other machines.
Thus the sign of the value returned when one of the
characters has its high-order bit set is
implementation-dependent.
.LP
Character movement is performed differently
in different implementations.
Thus overlapping moves may yield surprises.


diagnostics/fparel/memory.c
/*
 *      static char     fpasccsid[] = "@(#)memory.c 1.1 7/30/92 Sun right Sun Microsystems";
 */
#include <sys/types.h>
#include "fpa.h"


map_ram()
{
	u_long i, temp_val, tmp_data;
	u_long *ram_ptr, *rw_ram_ptr;
	u_char no_data_flag = 0xff; 

	ram_ptr = (u_long *)FPA_RAM_ACC;
	rw_ram_ptr = (u_long *)FPA_RW_RAM;
	*ram_ptr = TEST_PTR(0, MR_LB);
	temp_val = *rw_ram_ptr;
	temp_val = temp_val & MR_DMASK;
	if (temp_val)
		no_data_flag = 0x0; /* data  is non zero */
        for (i=1 ; i < MAXSIZE_MAP_RAM ; i++) {
                *ram_ptr = TEST_PTR(i, MR_LB);
                tmp_data = *rw_ram_ptr & MR_DMASK;
		if (tmp_data)
			no_data_flag = 0x0; /* data is non zero */
		temp_val = (temp_val ^ tmp_data) & MR_DMASK; 
        }
	if (no_data_flag)
		return(-1); /* data is zero , error */
	if (temp_val != 0xadface) /* the checksum should be ADFACE */
		return(-1);
	return(0);
}

ustore_ram()
{
        u_long i, tmp_val, tmp_data; 
	u_long tmp_val1, tmp_data1;
	u_long tmp_val2, tmp_data2;
	u_char no_data_flag = 0xff; 
	u_long *ram_ptr, *rw_ram_ptr; 

        ram_ptr = (u_long *)FPA_RAM_ACC; 
        rw_ram_ptr = (u_long *)FPA_RW_RAM; 
	 
	*ram_ptr = TEST_PTR(0, UR_LBL); 
        tmp_val = *rw_ram_ptr; 
        *ram_ptr = TEST_PTR(0, UR_LBM); 
        tmp_val1 = *rw_ram_ptr; 
        *ram_ptr = TEST_PTR(0, UR_LBH); 
        tmp_val2 = *rw_ram_ptr & UR_DMASK; 
	if ((tmp_val) || (tmp_val1) || (tmp_val2))
		no_data_flag = 0x0; /* data is non zero */

        for (i=1 ; i < MAXSIZE_USTORE_RAM ; i++) {

                *ram_ptr = TEST_PTR(i, UR_LBL);
                tmp_data = *rw_ram_ptr;
                *ram_ptr = TEST_PTR(i, UR_LBM);
                tmp_data1 = *rw_ram_ptr;
                *ram_ptr = TEST_PTR(i, UR_LBH);
                tmp_data2 = *rw_ram_ptr & UR_DMASK;
		if ((tmp_data) || (tmp_data1) || (tmp_data2))
			no_data_flag = 0x0;  /* data is non zero */

		tmp_val = tmp_val ^ tmp_data;
		tmp_val1 = tmp_val1 ^ tmp_data1;
		tmp_val2 = (tmp_val2 ^ tmp_data2) & UR_DMASK;
	}
	if (no_data_flag)
		return(-1); /* data is zero , error */
	if ((tmp_val == 0xBEADFACE) && (tmp_val1 == 0xBEADFACE) && (tmp_val2 == 0xCE)) 
			
			return(0);
	
	else
		return(-1);
}

reg_uh_ram()
{
        u_long i, tmp_val, tmp_data;  
        u_long tmp_val1, tmp_data1; 
        u_long *ram_ptr, *rw_ram_ptr;  
	u_char no_data_flag = 0xff;

	ram_ptr = (u_long *)FPA_RAM_ACC;
        rw_ram_ptr = (u_long *)FPA_RW_LREG_RR;
	*ram_ptr = TEST_PTR(0x400, RR_LB); 
        tmp_val = *rw_ram_ptr;
	if (tmp_val)
		no_data_flag = 0x0; /* data is non zero */

        for (i=0x401; i <= MAXSIZE_CON_RAM ; i++) {

                *ram_ptr = TEST_PTR(i, RR_LB);
                tmp_data = *rw_ram_ptr;
		if (tmp_data) 
	                no_data_flag = 0x0; /* data is non zero */
		tmp_val = tmp_val ^ tmp_data;
        }
	*ram_ptr = TEST_PTR(0x7FF,RR_LB);
	tmp_data = *rw_ram_ptr;
	if (tmp_data)
                no_data_flag = 0x0; /* data is non zero */
	tmp_val = tmp_val ^ tmp_data;

        rw_ram_ptr = (u_long *)FPA_RW_MREG_RR;
        *ram_ptr = TEST_PTR(0x400, RR_LB);  
        tmp_val1 = *rw_ram_ptr; 
	if (tmp_val1) 
                no_data_flag = 0x0; /* data is non zero */

        for (i=0x401; i <= MAXSIZE_CON_RAM ; i++) {

                *ram_ptr = TEST_PTR(i, RR_LB);
		tmp_data1 = *rw_ram_ptr; 
		if (tmp_data1) 
	                no_data_flag = 0x0; /* data is non zero */
                tmp_val1 = tmp_val1 ^ tmp_data1;
        }
	*ram_ptr = TEST_PTR(0x7FF, RR_LB);
        tmp_data1 = *rw_ram_ptr;
        if (tmp_data1)    
                no_data_flag = 0x0; /* data is non zero */
	tmp_val1 = tmp_val1 ^ tmp_data1;

	if (no_data_flag) {
		return(-1); /* data is zero , error */	
	}
	if ((tmp_val == 0xBEADFACE) && (tmp_val1 == 0xBEADFACE))
		
			return(0);
        else 
                return(-1); 
	
} 
 
 
